src/PermissionTree.js

```diff
@@ -0,0 +1,167 @@
+        }
+
+        if (!this.userPermissionTrees.has(user.id)) {
+            // Deep clone
+            this.setUserPermissionTree(user, JSON.parse(JSON.stringify(this.getDefaultTree())));
+        }
+
+        return this.userPermissionTrees.get(user.id);
+    }
+
+    setUserPermissionTree(user, tree) {
+        this.userPermissionTrees.set(user.id, tree);
+    }
+
+
+    buildACLQueries(roleName) {
+        const promises = [];
+        const defaultTree = this.getDefaultTree();
+
+        Object.keys(defaultTree).forEach((modelName) => {
+            Object.keys(defaultTree[modelName]).forEach((remoteMethod) => {
+                Object.keys(defaultTree[modelName][remoteMethod]).forEach((accessType) => {
+                    promises.push(new Promise((resolve, reject) => {
+                        this.models.ACL.checkPermission(
+                            this.models.RoleMapping.ROLE,
+                            roleName,
+                            modelName,
+                            remoteMethod,
+                            accessType,
+                            (err, accessRequest) => {
+                                if (err) reject(err);
+
+                                resolve(accessRequest);
+                            },
+                        );
+                    }));
+                });
+            });
+        });
+
+
+        return promises;
+    }
+
+    async getACLPermissionsForRole(role) {
+        return Promise.all(this.buildACLQueries(role));
+    }
+
+    async createPermissionTree(user) {
+        if (!user || !user.userGroups) {
+            throw new Error('No user was specified while trying to load permission tree!');
+        }
+
+        // Note: Roles are passed as user groups (RoleMapping.GROUP)
+        await Promise.all(user.userGroups.map(async(group) => {
+            const accessRequests = await this.getACLPermissionsForRole(group);
+
+            this.updatePermissions(user, accessRequests);
+        }));
+    }
+
+    updatePermissions(user, accessRequests) {
+        accessRequests.forEach((accessRequest) => {
+            if (
+                !this.getPermission(
+                    user,
+                    accessRequest
+                ) &&
+                accessRequest.isAllowed()
+            ) {
+                this.setPermission(
+                    user,
+                    accessRequest,
+                    true
+                );
+            }
+        });
+    }
+
+    getPermission(user, accessRequest) {
+        const userTree = this.getUserPermissionTree(user);
+
+        return userTree[accessRequest.model][accessRequest.property][accessRequest.accessType];
+    }
+
+    setPermission(user, accessRequest, allow = true) {
+        const userTree = this.getUserPermissionTree(user);
+
+        userTree[accessRequest.model][accessRequest.property][accessRequest.accessType] = allow;
+
+        this.setUserPermissionTree(user, userTree);
```

Comment:
You are modifying the `userTree` by reference, so, is there a reason you set the permission tree again?

Reply:
Makes sense!